scio-core/src/main/java/com/spotify/scio/transforms/BatchedBaseAsyncLookupDoFn.java

```diff
@@ -0,0 +1,325 @@
+  private final ConcurrentMap<String, Triple<Input, Instant, BoundedWindow>> inputsMap =
+      new ConcurrentHashMap<>();
+  private final ConcurrentLinkedQueue<Result> results = new ConcurrentLinkedQueue<>();
+  private final ConcurrentLinkedQueue<Pair<UUID, Input>> inputElements =
+      new ConcurrentLinkedQueue<>();
+  private long requestCount;
+  private long resultCount;
+  private final SerializableFunction<List<Input>, BatchRequest> batchRequestFn;
+  private final SerializableFunction<BatchResponse, List<Pair<String, Output>>> batchResponseFn;
+  private final SerializableFunction<Input, String> idExtractorFn;
+  private final CacheSupplier<String, Output> cacheSupplier;
+
+  public BatchedBaseAsyncLookupDoFn(
+      int batchSize,
+      int maxPendingRequests,
+      CacheSupplier<String, Output> cacheSupplier,
+      SerializableFunction<List<Input>, BatchRequest> batchRequestFn,
+      SerializableFunction<BatchResponse, List<Pair<String, Output>>> batchResponseFn,
+      SerializableFunction<Input, String> idExtractorFn) {
+    this.cacheSupplier = cacheSupplier;
+    this.batchRequestFn = batchRequestFn;
+    this.batchResponseFn = batchResponseFn;
+    this.idExtractorFn = idExtractorFn;
+    this.maxPendingRequests = maxPendingRequests;
+    this.batchSize = batchSize;
+    this.semaphore = new Semaphore(maxPendingRequests);
+  }
+
+  protected abstract ClientType newClient();
+
+  public abstract FutureType asyncLookup(ClientType client, BatchRequest input);
+
+  public abstract TryWrapper success(Output output);
+
+  public abstract TryWrapper failure(Throwable throwable);
+
+  @Override
+  public Pair<ClientType, Cache<String, Output>> createResource() {
+    return Pair.of(newClient(), cacheSupplier.get());
+  }
+
+  public ClientType getResourceClient() {
+    return getResource().getLeft();
+  }
+
+  public Cache<String, Output> getResourceCache() {
+    return getResource().getRight();
+  }
+
+  @StartBundle
+  public void startBundle(StartBundleContext context) {
+    futures.clear();
+    results.clear();
+    inputsMap.clear();
+    requestCount = 0;
+    resultCount = 0;
+    semaphore.drainPermits();
+    semaphore.release(maxPendingRequests);
+  }
+
+  @ProcessElement
+  public void processElement(
+      @Element Input input,
+      @Timestamp Instant timestamp,
+      OutputReceiver<KV<Input, TryWrapper>> out,
+      BoundedWindow window) {
+    flush(r -> out.output(KV.of(r.input, r.output)));
+    final Cache<String, Output> cache = getResourceCache();
+
+    try {
+
+      final Output cached = cache.getIfPresent(input);
+
+      if (cached != null) {
+        // found in cache
+        out.output(KV.of(input, success(cached)));
+      } else {
+        final UUID uuid = UUID.randomUUID();
+        inputElements.offer(Pair.of(uuid, input));
+        inputsMap.put(this.idExtractorFn.apply(input), Triple.of(input, timestamp, window));
```

Comment:
I think we can have issue here when 2 same ids are in the same batch. We'll loose the timestamp and window info from the 1st element.